src/coreclr/jit/compiler.h

```diff
@@ -7412,6 +7412,17 @@ class Compiler
         optNoReturnCallCount++;
     }
 
+    void setCallDoesNotReturn(GenTreeCall* const call)
+    {
+        assert(call != nullptr);
+        assert(!call->IsNoReturn());
+        call->gtCallMoreFlags |= GTF_CALL_M_DOES_NOT_RETURN;
+
+        // Always increment the root compiler's optNoReturnCallCount
+        // so fgTailMergeThrows has the correct number of candidates.
+        impInlineRoot()->setMethodHasNoReturnCalls();
```

Comment:
Is it ok that we might increment it here for an inlinee, but then reject that inlinee during importer? so we'll have more than needed noReturnCalls in the root

Reply:
That sounds reasonable to me. Do we already reject inlinees that throw?

Reply:
Inliner has two modes: 
1) static analysis
2) let's opportunistically start inlining and if we meet a patter we can't inline with - reject

for the 2nd case it could be e.g. a static field load that requires cctor init.

I don't see what protects you from the 2nd case here. We generally avoid changing root compiler's properties like this unless they're opportunistic and it's fine for them to be false-positive

Reply:
A better approach is to
1) Either calculate all no-returns in some phase after inlining is done
2) Or increment non-root compiler's counter. Then, there is a code path where we copy inlinee's properties to root after inlining is finalized.

Reply:
@amanasifkhalid Actually, never mind,  after looking into code it looks like it's fine for this case to be false positive